### DISCUSSION TITLE
What is the state of registers on reset?

### DISCUSSION THREAD

**Discussion by Azmisov** [SpenceKonde/megaTinyCore#827] [Q&A]
I've been looking around but can't seem to find this info in the datasheets or online. What can I assume about the values of registers when it first starts up (Attiny424 if it matters)? Are they always zeroed out, or undefined values? E.g. say I don't want interrupts enabled for timer A, do I need to explicitly zero out `TCA0.SPLIT.INTCTL`?

**Comment by jvasileff** [comment]
Looks like all zeros on reset for that one, no? Some others are documented to have initial values based on configuration, eg BOD.CTRLB is loaded from FUSE.BODCFG (18.5.2).

![E0A18ADD-68D3-4224-BED6-B115C4321FBD](https://user-images.githubusercontent.com/236050/191110002-594009d2-b34f-44d9-b8eb-877742b63363.png)

**Reply by Azmisov** [reply]
Oh I see, thanks. I didn't notice they have the "Reset" state listed right there.

**Comment by SpenceKonde** [comment]
All special function registers (the things we normally call registers unless we're talking about code written in assembly) are reset to a defined initial value. Code all over the cores assumes that registers that the core hasn;t changed are in their reset state (that's the origin of the warning that if you touch registers directly, you may or may not get expected behavior when you try to use Arduino API functions that work with it. That happens before the start of any code execution, during the hardware reset.
After that it starts executing code from 0x0000, which is a jump to the initialization code:
* static and global variables with non-zero values (.bss) are initialized
* r1 is set to 0 and the stack pointer is redundantly set up (because that happens automatically on modern AVRs,. It didn;t on classic ones as I understand it)
* It then reaches .init3 which is the first custom megaTinyCore code. This is where it checks and clears the reset flag register, It there was a reset flag set, it writes that value to GPIOR0 so the user can fish it out later without difficulty (and it can be freely written over, It's only important between this point in time, and when user code that needs to know what caused the last reset runs. After that, you can do whatever you want with GPIOR0, and if you never care about the reset flag, just be sure not to assume GPIOR is 0.)
  * If we do NOT see a reset flag, that means we arrived at 0x0000 WITHOUT resetting. That's a Bad Thing, we call it a dirty reset - that means none of the registers are reset to their default values because there was no reset. This can happen if you manually jump to 0x00, or more likely, smash the stack during a function call or interrupt and then return from it, ending up and a deterministic but unknown address in the flash. If you are lucky, this is blank flash after the end of the application, it skids along the 0xFFFF instructions (which do the same thing as 0xFFF7 - skip next instruction if bit 7 in working register r31 is set. It gets to the end of the flash, and then either executes 0x0000, or skips that and hits the nmi vector, but the nmi vector is never used in arduino, so it points to BADISR, which jumps to 0x0000. 
  * As you can see from that last sentence, an interrupt routine that does not have an ISR defined will end up jumping to BADISR, which jumps to 0x0000 in a dirty reset, and hence will trigger a hardware reset. 
* After that we go through things like running class constructors and shit, and then at the end of .init9. we jump to main(). main then calls the internal init() function which:
  * Sets the clock divider to get the clock speed that the user requested. We can't change the fuse to choose between 16 and 20 MHz base frequencies.... so you just sorta hope and pray that the user isn't using optiboot written with 16 MHz selected while the speed is now set to 20 MHz or vise versa.   
   * This uses a very different code path is tuned speeds are used, because there we expect that the chip has been tuned with either approximate (via self-tuning) or exact (via tuning with an actual timerback) tuning values and that those are stored in the last 12 bytes of the user row; So we check first the "home" speed (the 16 or 20 MHz "tuned" speed. If that's 0xFF, the chip must not have been tuned for this oscillator base frequency, so we take an empirical guess based on default cal to get desired speed (accuracy is dubious, but closer than you might think - one of my dreams is to gather the calibration vs F_CPU for a large enough sample that I can distill the curve to a single parameter, such that one could just look at the factory cal, and discover the calibration value for any desired speed without any tuning at all). After this, we go through the same clock division setting logic as we would have above (after setting the oscillator cal value). See Ref_Tuning.md
  * Turn on ADC0 (ADC1, if present, needs to be manually turned on, see Ref_Analog.md)
  * Set up TCA0 for split mode, and set the portmux to match the pins we decided to put PWM on (there are 6 channels, each gets 2 options). If it's a Dx-series impacted by the errata, we set ADC.MUXPOS = 0x40 (ground) so that it doesn't disable input on a pin.
  * If TCD is used for PWM but not millis (1-series with 20/24 pins but non default millis timer selection), we set that up for PWM. 
  * Set up millis() timekeeping based on the selected timer
  * The very last step between init() and setup() is the the sei to enable interrupts..
* main then calls setup()
* and ends in a while(1) loop(); to run loop forever.

At a few points in there there are weakly defined functions with empty function bodies that the user can override if they have to execute code before initialization, for exotic and advanced use cases. See Ref_Callbacks.

All the Ref_ documents can be found in https://github.com/SpenceKonde/megaTinyCore/tree/master/megaavr/extras

**RELATED TO YOUR SPECIFC CASE**
If TCA is selected for millis timekeeping, you must have an interrupt on TCA's underflow, Otherwise millis() will always report 0, while micros will count up to 815 and then roll over to 0. 
analogWrite() and digitalWrite() assume that the PWM-generating timers are configured how it set them in init(). If you're changing any setting on a PWM-generating timer, analogWrite() and digitalWrite() will still behave as if threre is PWM on those pins, 
What does that mean? 
That means that if you fiddle with those registers millis and/or PWM will break. Manually adjust a different timer, or put millis on a different timer. This is only relevant for TCA or TCD as millis timer (used as default on 0 and 1-series respectively. TCBs are MUCH better as a millis timer, but unfortunately, they're also the only game in town for servo and tone(), while TCA's can be used for millis and PWM, so that's why we did that on 0-series. 

On 1-series, there's a TCD that we use by default for millis. (we don't enable for PWM (yet) on WOA or WOB, because they overlap with pins used by TCA) This in spite of TCD0 being the worst timer for millis - it's also the most complicated peripheral on the part (longest chapter of the datasheet, and the datasheet is still terse in places leaving one with many question about how to use it, though, and the logic was that ain't nobody gonna figure that shit out, might as well use it for millis (and it can also do millis and PWM together). 

So **thou shalt not fiddle with the registers of the timer used for millis**

TCBs can do PWM, but they're shitty at it and only one TCB on the largest parts actually has a pin output. On 2-series parts, TCB1 is used for millis by default, leaving TCB1 for Servo or tone(). 

On a part with TCA0, TCB0 and TCB1 (tinyAVR 1 w/16 or 32k flash, and tinyAVR2) if TCA is used as the time, you should be able to use servo and tone simultaneously. 

If you are using the timer without actually outputting to pins (for instance, I do this on a t424 on my WS2812B signal router, which lets me use less than 1 microcontroller per string of lights when the lights aren't daisychainable or we don't want them daisychained because, maybe we want the lights in a coming out along lines from a center point - it takes 1 TCB, the TCA, and all 4 CCLs , leaving justenough pins for a TX_ONLY debug serial line, and a "sense line" for each connected string - we turn on the pullup, and on the light string, there's a resistor in the tens to hundreds of k between the sense pin and ground, and I only needed a few buckets to sort string lengths into (it,s either 50, 100, 200, or 332 for the important cases for me)

**Reply by SpenceKonde** [reply]
If you need or want to tell the core to leave TCD0 or TCA0 alone (so analogWrite, digitalWrite and turnOffPWM don't impact it), you can call takeOverTCA0() or takeOverTCD0() - these will mark those timers as off limits to the core. In the case of TCA0, it will also reset all the hardware registers to their default state using the hard reset command. No such command exists for the TCD. There a resumeTCA0() to pass controll of TCA0 back to the core so it works normally again. TCD on the other hand is a fiendishly complex device with no hard reset command, it is your responsibility to make sure the registers are set up the way you want there. 

See Ref_Serial for an example of bulk-printing the contents of a peripherals registers using the Serial.printHex() functions (my frequent need to dump the entire contents of a peripheral register the main reason I did that,though it has many uses; )

**Reply by Azmisov** [reply]
Okay thanks for the extremely detailed response, this is a really helpful reference.
